File: experiments/verification-poc/evidence.go
package main

import (
    "fmt"
    "github.com/SCKelemen/oak/experiments/verification-poc/frontend"
    "github.com/SCKelemen/oak/experiments/verification-poc/internal/lrat"
)
const evidenceFormat="oak-evidence-3"

// Verdict is the evidence contract (roadmap step 4): every accepted result
// states exactly what was established, under which assumptions, and through
// which trust path. Nothing in a verdict is inferred from the evidence itself;
// each field is fixed by the kind of evidence checked and by the checker's own
// recorded trust boundary (README, "Go packages and trust").
type Verdict struct {
    Accepted bool `json:"accepted"`
    // Claim is the property the evidence supports, about the identified
    // Oak declarations (roadmap step 3): the claim names the state record and
    // the initial, step, and invariant predicates by name and position in the
    // source the verdict binds by hash and semantic digest.
    Claim Claim `json:"claim"`
    // Established is the fact this checker verified about the evidence.
    Established string `json:"established"`
    // Assumptions are trusted, not checked, by this verdict.
    Assumptions []string `json:"assumptions"`
    // TrustPath lists the components the result passed through, in order.
    TrustPath []string `json:"trust_path"`
    // Unsupported names what the checker rejects rather than trusts.
    Unsupported []string `json:"unsupported,omitempty"`
    // Details carries the evidence-specific numbers and sub-results.
    Details map[string]any `json:"details,omitempty"`
}

// Claim is a property about identified Oak source declarations.
type Claim struct {
    Property string `json:"property"`
    Source string `json:"source"`
    SourceHash string `json:"source_sha256"`
    SemanticDigest string `json:"semantic_digest"`
    State frontend.Declaration `json:"state"`
    Initial frontend.Declaration `json:"initial"`
    Step frontend.Declaration `json:"step"`
    Invariant frontend.Declaration `json:"invariant"`
}

// claimFor names the declarations a property is about. A declaration the
// frontend did not locate keeps its name with no position.
func claimFor(m *Model,property string) Claim {
    locate:=func(name string) frontend.Declaration {
        if d,ok:=m.Declarations[name];ok{return d}
        return frontend.Declaration{Name:name}
    }
    return Claim{
        Property:property,
        Source:m.Config.Source,
        SourceHash:m.Document.SourceHash,
        SemanticDigest:m.Digest,
        State:locate(m.Document.State),
        Initial:locate(m.Config.Initial),
        Step:locate(m.Config.Step),
        Invariant:locate(m.Config.Invariant),
    }
}

// The recorded trust boundary shared by every verdict: the model is exported
// by Oak's frontend and evaluated by the Go typed-model adapter; the source
// and settings are bound by the semantic digest, not re-derived.
var sharedAssumptions=[]string{
    "Oak's frontend exports the narrow model faithfully from the checked program",
    "the Go typed-model adapter (truth, states, valid) implements the model's semantics",
    "the source bytes and project settings bound by the semantic digest are the ones reviewed",
    "Go and the execution environment",
}
var sharedTrustPath=[]string{
    "semantic digest binding (" + evidenceFormat + ")",
    "frontend/ model export",
}

func closedSetVerdict(m *Model,states int) *Verdict {
    return &Verdict{
        Accepted:true,
        Claim:claimFor(m,"nonvacuous reachable safety"),
        Established:"a finite set of states that contains every initial state (at least one), is closed under the step relation, and satisfies the invariant in every member",
        Assumptions:sharedAssumptions,
        TrustPath:append(append([]string{},sharedTrustPath...),"Go closed-set checker (evidence.go)"),
        Details:map[string]any{"states":states},
    }
}

func traceVerdict(m *Model,states int) *Verdict {
    return &Verdict{
        Accepted:true,
        Claim:claimFor(m,"reachable safety counterexample"),
        Established:"a trace that starts in an initial state, follows legal transitions or stutters, and ends in a state violating the invariant",
        Assumptions:sharedAssumptions,
        TrustPath:append(append([]string{},sharedTrustPath...),"Go trace replay (evidence.go)"),
        Details:map[string]any{"states":states},
    }
}

func lratVerdict(m *Model,checks map[string]lrat.Result) *Verdict {
    return &Verdict{
        Accepted:true,
        Claim:claimFor(m,"nonvacuous inductive safety"),
        Established:"an initial-state witness, and LRAT refutations of the initialization and preservation counterexample formulas reconstructed from the source (their SHA-256 matched)",
        Assumptions:append(append([]string{},sharedAssumptions...),
            "the Go CNF translation (encode) is faithful to the model; it is compared, not proved"),
        TrustPath:append(append([]string{},sharedTrustPath...),
            "Go CNF encoder (circuit.go)",
            "internal/lrat checker: its decisions agree with compiled Oak rup_text_check and the Lean file model CertificateFile.check on every corpus case; the Lean model is proved sound (check_sound) and the transliterated Oak decoder proved to refine it (OakTextRefinement.check_refines)"),
        Unsupported:[]string{"RAT hints","binary LRAT","extension variables","SMT proof formats"},
        Details:map[string]any{"obligations":checks},
    }
}
type Proof struct{ SHA string `json:"cnf_sha256"`; LRAT string `json:"lrat"` }
type Certificate struct{
    Format string `json:"format"`
    Digest string `json:"semantic_digest"`
    Kind string `json:"kind"`
    Initial State `json:"initial_witness,omitempty"`
    Proofs map[string]Proof `json:"proofs,omitempty"`
    States []State `json:"states,omitempty"`
}
func verify(m *Model,c *Certificate)(*Verdict,error){
    if c.Format!=evidenceFormat||c.Digest!=m.Digest{return nil,fmt.Errorf("wrong evidence format or source/model identity")}
    if c.Kind=="closed-set" {
        if c.Initial!=nil||c.Proofs!=nil||len(c.States)==0||len(c.States)>1024{return nil,fmt.Errorf("invalid closed-set evidence")}
        members:=map[string]bool{}
        for _,s:=range c.States{if e:=m.valid(s);e!=nil{return nil,e};key:=stateKey(s);if members[key]{return nil,fmt.Errorf("duplicate closed-set state")};members[key]=true;if !truth(m.Terms["invariant"],s,nil){return nil,fmt.Errorf("unsafe closed-set member")}}
        all,e:=m.states();if e!=nil{return nil,e};initials:=0
        for _,s:=range all{if truth(m.Terms["initial"],s,nil){initials++;if !members[stateKey(s)]{return nil,fmt.Errorf("closed set omits an initial state")}}}
        if initials==0{return nil,fmt.Errorf("empty initial set")}
        for _,s:=range c.States{for _,target:=range all{if truth(m.Terms["step"],s,target)&&!members[stateKey(target)]{return nil,fmt.Errorf("closed set omits a successor")}}}
        return closedSetVerdict(m,len(c.States)),nil
    }
    if c.Kind=="trace"{
        if c.Initial!=nil||c.Proofs!=nil{return nil,fmt.Errorf("unexpected trace fields")}
        if e:=replay(m,c.States);e!=nil{return nil,e}
        if truth(m.Terms["invariant"],c.States[len(c.States)-1],nil){return nil,fmt.Errorf("trace does not end in safety violation")}
        return traceVerdict(m,len(c.States)),nil
    }
    if c.Kind!="lrat"||c.States!=nil||len(c.Proofs)!=2{return nil,fmt.Errorf("invalid proof evidence")}
    if e:=m.valid(c.Initial);e!=nil{return nil,e};if !truth(m.Terms["initial"],c.Initial,nil){return nil,fmt.Errorf("invalid initial witness")}
    checks:=map[string]lrat.Result{}
    for _,role:=range []string{"base","step"}{proof,ok:=c.Proofs[role];if !ok{return nil,fmt.Errorf("missing %s proof",role)};circuit,r:=encode(m,role);cnf:=circuit.dimacs(r);if proof.SHA!=digest(cnf){return nil,fmt.Errorf("proof CNF differs from reconstructed %s obligation",role)};result,e:=lrat.Check(cnf,proof.LRAT);if e!=nil{return nil,fmt.Errorf("%s: %w",role,e)};checks[role]=result}
    return lratVerdict(m,checks),nil
}
func replay(m *Model,states []State)error{
    if len(states)==0||len(states)>4096{return fmt.Errorf("invalid trace length")}
    for _,s:=range states{if e:=m.valid(s);e!=nil{return e}}
    if !truth(m.Terms["initial"],states[0],nil){return fmt.Errorf("invalid initial state")}
    for i:=1;i<len(states);i++{s,t:=states[i-1],states[i];if stateKey(s)!=stateKey(t)&&!truth(m.Terms["step"],s,t){return fmt.Errorf("illegal transition at %d",i)}};return nil
}

File: experiments/verification-poc/evidence.py
#!/usr/bin/env python3
"""Independent evidence acceptance. Does not import or run the search engine."""
import argparse
import itertools
import json
from pathlib import Path
import sys
from project import load_project

FORMAT = 'oak-evidence-1'


class Rejected(ValueError):
    pass


def require(condition, reason):
    if not condition:
        raise Rejected(reason)


def exact(value, keys):
    require(type(value) is dict and set(value) == set(keys), 'unexpected or missing evidence fields')


def read_json(path):
    def pairs(items):
        result = {}
        for key, value in items:
            require(key not in result, 'duplicate JSON key')
            result[key] = value
        return result
    require(path.stat().st_size <= 2_000_000, 'evidence exceeds prototype size limit')
    return json.loads(path.read_text(), object_pairs_hook=pairs)


def state(value, fields):
    exact(value, fields)
    require(all(type(v) is bool for v in value.values()), 'state fields must be actual Booleans')
    return tuple(value[f] for f in fields)


def truth(expr, assignment):
    """Three-valued partial evaluation; None means not established."""
    op = expr[0]
    if op == 'bool':
        return expr[1]
    if op == 'var':
        return assignment.get(expr[1] + '.' + expr[2])
    a = truth(expr[1], assignment)
    if op == '!':
        return None if a is None else not a
    b = truth(expr[2], assignment)
    if op == '&&':
        if a is False or b is False:
            return False
        return True if a is True and b is True else None
    if op == '||':
        if a is True or b is True:
            return True
        return False if a is False and b is False else None
    require(op in ('==', '!='), 'unsupported logical operator')
    if a is None or b is None:
        return None
    return (a == b) if op == '==' else (a != b)


def assignment(**states):
    return {p + '.' + f: v for p, s in states.items() for f, v in s.items()}


def instantiate(expr, old, new):
    if expr[0] == 'var':
        return ('var', new if expr[1] == old else expr[1], expr[2])
    if expr[0] == 'bool':
        return expr
    return (expr[0], *(instantiate(x, old, new) for x in expr[1:]))


def obligations(terms):
    return {
        'base': ('&&', terms['initial'], ('!', terms['invariant'])),
        'step': ('&&', ('&&', terms['invariant'], terms['step']),
                 ('!', instantiate(terms['invariant'], 's', 't'))),
    }


def check_tree(tree, formula, allowed, env=None):
    """Rules: false leaf; or both Boolean cases of a fresh variable."""
    env = {} if env is None else env
    require(type(tree) is dict, 'proof node must be an object')
    if set(tree) == {'false'}:
        require(tree['false'] is True, 'false leaf must have literal true marker')
        require(truth(formula, env) is False, 'leaf does not establish falsity')
        return 1
    exact(tree, ('split', 'zero', 'one'))
    variable = tree['split']
    require(type(variable) is str and variable in allowed, 'unknown split variable')
    require(variable not in env, 'repeated split variable')
    # Both branches mandatory, and their assignments cannot be supplied by evidence.
    return 1 + check_tree(tree['zero'], formula, allowed, env | {variable: False}) + check_tree(tree['one'], formula, allowed, env | {variable: True})


def verify(project, evidence):
    model, terms, digest = load_project(project)
    require(type(evidence) is dict, 'evidence must be an object')
    require(evidence.get('format') == FORMAT, 'unsupported format')
    require(evidence.get('semantic_digest') == digest, 'evidence belongs to a different source/project/version')
    kind = evidence.get('kind')
    common = ('format', 'semantic_digest', 'kind')
    fields = model.fields
    if kind == 'trace':
        exact(evidence, (*common, 'states'))
        states = evidence['states']
        require(type(states) is list and 1 <= len(states) <= 4096, 'invalid trace length')
        for s in states:
            state(s, fields)
        require(truth(terms['initial'], assignment(s=states[0])) is True, 'trace does not start in an initial state')
        for s, t in zip(states, states[1:]):
            require(s == t or truth(terms['step'], assignment(s=s, t=t)) is True, 'trace contains an illegal transition')
        require(truth(terms['invariant'], assignment(s=states[-1])) is False, 'trace does not end in a safety violation')
        return {'accepted': True, 'claim': 'reachable safety counterexample', 'states': len(states)}
    if kind == 'inductive':
        exact(evidence, (*common, 'initial_witness', 'proofs'))
        state(evidence['initial_witness'], fields)
        require(truth(terms['initial'], assignment(s=evidence['initial_witness'])) is True, 'invalid initial witness')
        exact(evidence['proofs'], ('base', 'step'))
        counts = {}
        for name, formula in obligations(terms).items():
            allowed = {p + '.' + f for p in (('s',) if name == 'base' else ('s', 't')) for f in fields}
            counts[name] = check_tree(evidence['proofs'][name], formula, allowed)
        return {'accepted': True, 'claim': 'nonvacuous inductive safety', 'proof_nodes': counts}
    require(kind == 'closed-set', 'unsupported evidence kind')
    exact(evidence, (*common, 'states'))
    supplied = evidence['states']
    require(type(supplied) is list and 1 <= len(supplied) <= 2 ** len(fields), 'invalid closed-set size')
    keys = [state(s, fields) for s in supplied]
    require(len(set(keys)) == len(keys), 'duplicate states')
    universe = [dict(zip(fields, bits)) for bits in itertools.product((False, True), repeat=len(fields))]
    initial = [s for s in universe if truth(terms['initial'], assignment(s=s)) is True]
    require(bool(initial), 'empty initial set')
    require(all(state(s, fields) in keys for s in initial), 'certificate omits an initial state')
    for s in supplied:
        require(truth(terms['invariant'], assignment(s=s)) is True, 'certificate includes unsafe state')
        for t in universe:
            if truth(terms['step'], assignment(s=s, t=t)) is True:
                require(state(t, fields) in keys, 'certificate is not closed under transitions')
    return {'accepted': True, 'claim': 'finite-state safety via closed invariant set', 'states': len(keys)}


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('project', type=Path)
    parser.add_argument('evidence', type=Path)
    args = parser.parse_args()
    try:
        result = verify(args.project, read_json(args.evidence))
        print(json.dumps(result, indent=2))
        return 0
    except (ValueError, OSError, RecursionError, TypeError) as error:
        print(json.dumps({'accepted': False, 'reason': str(error)}))
        return 1


if __name__ == '__main__':
    sys.exit(main())
